src/poetry/utils/cache.py

```diff
@@ -0,0 +1,250 @@
+from __future__ import annotations
+
+import contextlib
+import hashlib
+import json
+import shutil
+import time
+
+from dataclasses import dataclass
+from dataclasses import field
+from pathlib import Path
+from typing import Any
+from typing import Callable
+from typing import Generic
+from typing import TypeVar
+
+
+T = TypeVar("T")
+
+
+def decode(string: bytes, encodings: list[str] | None = None) -> str:
+    """
+    Compatiblity decode function pulled from cachy.
+
+    :param string: The byte string to decode.
+    :param encodings: List of encodings to apply
+    :return: Decoded string
+    """
+    if encodings is None:
+        encodings = ["utf-8", "latin1", "ascii"]
+
+    for encoding in encodings:
+        with contextlib.suppress(UnicodeDecodeError):
+            return string.decode(encoding)
+
+    return string.decode(encodings[0], errors="ignore")
+
+
+def encode(string: str, encodings: list[str] | None = None) -> bytes:
+    """
+    Compatibility encode function from cachy.
+
+    :param string: The string to encode.
+    :param encodings: List of encodings to apply
+    :return: Encoded byte string
+    """
+    if encodings is None:
+        encodings = ["utf-8", "latin1", "ascii"]
+
+    for encoding in encodings:
+        with contextlib.suppress(UnicodeDecodeError):
+            return string.encode(encoding)
+
+    return string.encode(encodings[0], errors="ignore")
+
+
+def _expiration(minutes: int) -> int:
+    """
+    Calculates the time in seconds since epoch that occurs 'minutes' from now.
+
+    :param minutes: The number of minutes to count forward
+    """
+    return round(time.time()) + minutes * 60
+
+
+@dataclass
+class Store(Generic[T]):
+    """
+    Abstract class for representing a cache store.
+    """
+
+    def get(self, key: str) -> T | None:
+        """
+        Retrieve an item from the cache by key.
+
+        :param key: The cache key
+
+        :return: The cache value
+        """
+        raise NotImplementedError()
+
+    def has(self, key: str) -> bool:
+        """
+        Determine if a file exists and has not expired in the cache.
+
+        :param key: The cache key
+
+        :returns: True if the key exists in the cache
+        """
+        return self.get(key) is not None
```

Comment:
Unsure if file stat is sufficient as file timestamps are not reliable across filesystems and/or cases when cache is restored under different environments (eg: CI).

Reply:
Right now the actual filesystem-based cache only saves entries indefinitely, so no timestamp checking is needed either. At this moment, it would be functionally equivalent to check the filesystem for the existence of this file.

This can be confirmed by checking caches directly as well:

    find ~/.cache/pypoetry/cache/repositories/pypi -not -type d -not -path "*/_http/*" -exec sh -c "echo -n '{} '; head -c 10 {}; echo" \;

The header of each file is the unix timestamp when the cache expires -- which is 9999999999 everywhere.